### PR TITLE
Add clock type to node_options

### PR DESCRIPTION
**Pull request by jefferyyjhsu**
Signed-off-by: Jeffery Hsu [jefferyyjhsu@gmail.com](mailto:jefferyyjhsu@gmail.com)
This PR adds the option for selecting clock source in Node/LifecycleNode thru NodeOptions.
A similar option is already present in rclcpp::Timer but it is currently missing in Nodes/LifecycleNode. There's currently no way to set Node::now() and all other time-related functions to use clock sources other than ROS_TIME.

### CODE REVIEW

**Comment by alsora** [comment]
CI
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/17136/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/11683/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/17554/)

**Review by alsora — APPROVED** [review]

**Comment by alsora** [comment]
NOTE: this PR is from iRobot.
It looks good to me, but I would like also feedbacks from other reviewers

File: rclcpp/src/rclcpp/node_interfaces/node_clock.cpp
Comment (fujitatomoya) [inline rclcpp/src/rclcpp/node_interfaces/node_clock.cpp:34]: would it be nice to change the name into `node_clock_` or something indicates that belongs to node? i am not sure why this is statically set with `RCL_ROS_TIME` though.

File: rclcpp/include/rclcpp/node_options.hpp
Comment (fujitatomoya) [inline rclcpp/include/rclcpp/node_options.hpp:262]: test would be ideal.

**Review by fujitatomoya — COMMENTED** [review]
lgtm with test for `NodeOption`

File: rclcpp/src/rclcpp/node_interfaces/node_clock.cpp
Comment (alsora) [reply rclcpp/src/rclcpp/node_interfaces/node_clock.cpp:34]: +1 on changing name as you suggested.

IMO that was just an oversight that no one tried to fix..
The ROS time is non monotonic, so it's a very dangerous default if your code expects time to be linear.

Comment (clalancette) [reply rclcpp/src/rclcpp/node_interfaces/node_clock.cpp:34]: > The ROS time is non monotonic, so it's a very dangerous default if your code expects time to be linear.

I mean, yes, but if you don't deal with it then your code has no chance at all to work with simulation.

It's not exactly clear to me what this PR is doing, but I think we absolutely should keep the default as RCL_ROS_TIME, otherwise simulation will just not work properly.  Further, since this PR is changing a public API (the NodeClock constructor), we should absolutely tick-tock the API change (assuming others disagree with me and think we should always require users to pass the clock type in).

Comment (alsora) [reply rclcpp/src/rclcpp/node_interfaces/node_clock.cpp:34]: This PR adds the possibility to select the clock type of a node via rclcpp::Node Options.

Otherwise APIs such as Node::now()  can't be configured to use mononotic time.

Note that this does not change the default (which remains ROS time).

Comment (jefferyyjhsu) [reply rclcpp/src/rclcpp/node_interfaces/node_clock.cpp]: > would it be nice to change the name into `node_clock_` or something indicates that belongs to node? i am not sure why this is statically set with `RCL_ROS_TIME` though.
I change the name to `clock_` since the class itself is named `NodeClock`.

**Comment by asymingt** [comment]
Forgive me if I misunderstand this PR, but I think being able to have `Node::now()` driven by sim time is currently possible with setting `use_sim_time=true` parameter on the node. After [this PR] lands you should be able to create both sim and wall timers with `Node::create_timer()` and `Node::create_wall_timer()`. Is there a use case beyond these that I'm missing?

**Comment by alsora** [comment]
@asymingt the purpose of this PR is to allow to set the node clock type.
There are 3 types of clocks in ros: real time, monotonic and "ros".

Currently, you must use a clock of type "ros" for a node.
A clock of this type uses a real time clock and also the clock topic if available (and if use_sim_time is set to true).

There are a lot of applications where a real time clock is not suitable and it is required to use a monotonic clock.

**Comment by asymingt** [comment]
> @asymingt the purpose of this PR is to allow to set the node clock type. There are 3 types of clocks in ros: real time, monotonic and "ros".
> 
> Currently, you must use a clock of type "ros" for a node. A clock of this type uses a real time clock and also the clock topic if available (and if use_sim_time is set to true).
> 
> There are a lot of applications where a real time clock is not suitable and it is required to use a monotonic clock (and as said before, the "ros" clock is not monotoni)

Understood. Thank you, and sorry for the distraction.

**Comment by alsora** [comment]
New CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/17194/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/11735/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/17616/)

**Comment by alsora** [comment]
@fujitatomoya I think that comments have been addressed.
Anything else to do?

File: rclcpp/test/rclcpp/test_node.cpp
Comment (fujitatomoya) [inline rclcpp/test/rclcpp/test_node.cpp:165]: So user must make sure that if they want to use ROS simulation time, the Node::Option clock_type must be ROS time statically?

What about the user changes `use_sim_time` parameter to `true` when they want to use simulation time for the node afterward? i think this is coming back to @clalancette 's comment https://github.com/ros2/rclcpp/pull/1982#discussion_r947205947?
I do not think this is gonna work with simulation time if user enables `use_sim_time` later after node initialization.

**Comment by fujitatomoya** [comment]
either @iuhilnehc-ynos or @Barry-Xu-2018 , could you help us review on this?

I was thinking that, i may be missing something.

- `TimeSource` and `ClockState` are dedicated to manage the ROS simulation time. User must be able to switch on/off simulation time afterward.
- Having `Node::Option::clock_type` can be used to default clock type creating timer and so on.

CC: @alsora @jefferyyjhsu

File: rclcpp/test/rclcpp/test_node.cpp
Comment (jefferyyjhsu) [reply rclcpp/test/rclcpp/test_node.cpp:165]: Since `on_parameter_event()` eventually calls `set_clock()` when `use_sim_time` is changed, it will cause an exception throw if the clock is not of `RCL_ROS_TIME` type. I believe it's the correct behavior as only `RCL_ROS_TIME` allows sim time.
source:
https://design.ros2.org/articles/clock_and_time.html#time-abstractions

Comment (iuhilnehc-ynos) [reply rclcpp/test/rclcpp/test_node.cpp:165]: `clocks_state_.enable_ros_time`  in `attachNode` throwing an exception makes sense, but is it good to throw an exception in `on_parameter_event`? I think using `RCLCPP_ERROR` in the `on_parameter_event` might be better.
It seems that [the logic of throwing the exception](https://github.com/ros2/rclcpp/blob/bca2d75fc4caa532b38580d0135b48c79e111874/rclcpp/src/rclcpp/time_source.cpp#L148) should move up.

File: rclcpp/src/rclcpp/node_options.cpp
Comment (iuhilnehc-ynos) [inline rclcpp/src/rclcpp/node_options.cpp:23]: remove this line because `node_options.hpp` already include it.

**Review by iuhilnehc-ynos — COMMENTED** [review]
minor nitpick

**Comment by alsora** [comment]
@fujitatomoya I agree on having the clock type to be used as the default time for clocks... I was assuming that it was already the case.

About simulated time: users should be able to turn it on/off UNLESS they manually set the node clock to monotonic... In that case simulated time is not currently expected to work and it should throw an exception/error somewhere.

**Comment by alsora** [comment]
CI
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/17208/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/11751/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/17641/)

**Comment by alsora** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/17220/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/11763/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/17680/)

**Comment by alsora** [comment]
Hi, are there any more changes that we should address before proceeding with this PR?

File: rclcpp/src/rclcpp/node_options.cpp
Comment (fujitatomoya) [reply rclcpp/src/rclcpp/node_options.cpp:23]: @jefferyyjhsu did you address this one?

File: rclcpp/test/rclcpp/test_node.cpp
Comment (fujitatomoya) [reply rclcpp/test/rclcpp/test_node.cpp:165]: > I believe it's the correct behavior as only RCL_ROS_TIME allows sim time.

i think this is correct. but related to previous @iuhilnehc-ynos comment, i do not think that is a good behavior.

> I think using RCLCPP_ERROR in the on_parameter_event might be better.

this could be one solution. but before that, probably it should not be able to change the parameter `use_sim_time` into `true` in this case? because we will throw the exception afterwards.
It would be better to reject the parameter change and let the user application know that `you cannot use simulation time, because node specifies to use system or monotonic clock type via option`?

**Comment by fujitatomoya** [comment]
@alsora @jefferyyjhsu sorry to be late, just left couple of comments.

File: rclcpp/src/rclcpp/node_options.cpp
Comment (jefferyyjhsu) [reply rclcpp/src/rclcpp/node_options.cpp]: Sorry! I missed this one earlier.

File: rclcpp/test/rclcpp/test_node.cpp
Comment (jefferyyjhsu) [reply rclcpp/test/rclcpp/test_node.cpp:165]: @fujitatomoya, I agree.
I added a new `on_set_parameters_callback` to reject `use_sim_time` parameter changes altogether if the clock_type is not compatible.

Comment (fujitatomoya) [reply rclcpp/test/rclcpp/test_node.cpp:165]: @iuhilnehc-ynos could you review on this too to share your comment?

**Comment by fujitatomoya** [comment]
@ivanpauno can you also take a look at this, this changes behavior a bit.

CC: @clalancette

File: rclcpp/src/rclcpp/time_source.cpp
Comment (alsora) [inline rclcpp/src/rclcpp/time_source.cpp:96]: we should replace `ros_time_active_ == true` with just `ros_time_active_`

File: rclcpp/src/rclcpp/time_source.cpp
Comment (alsora) [inline rclcpp/src/rclcpp/time_source.cpp:139]: minor: I think that the code is more readable if rather than having this `} else if (set_ros_time_enabled && !clock->ros_time_is_active()) {` and then `} else if (set_ros_time_enabled) {` we rather do

```
} else if (set_ros_time_enabled) {
  if (clock->ros_time_is_active()) {
    throw std::invalid_argument(....)
  }
  ....
}
```

File: rclcpp/src/rclcpp/time_source.cpp
Comment (alsora) [inline rclcpp/src/rclcpp/time_source.cpp:312]: it seems like other parts of this class are not checking for whether `node_parameters_` is a valid pointer or not

File: rclcpp/src/rclcpp/time_source.cpp
Comment (alsora) [inline rclcpp/src/rclcpp/time_source.cpp:458]: the `on_parameter_event` function below does
```
      if (it.second->value.type != ParameterType::PARAMETER_BOOL) {
        RCLCPP_ERROR(logger_, "use_sim_time parameter cannot be set to anything but a bool");
        continue;
      }
```

should maybe this check on the type be moved here? we should fail the "parameter set" if someone uses the wrong type.

**Comment by alsora** [comment]
Changes look good to me, I just left some minor comments.
New CI

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/17374/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/11908/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/17876/)

File: rclcpp/src/rclcpp/time_source.cpp
Comment (jefferyyjhsu) [reply rclcpp/src/rclcpp/time_source.cpp:312]: True. I was following how it's done in 9723576821304add8993c7e9a3057fa284c1146f, but I agree that we should add the check in `attachNode()`.

File: rclcpp/src/rclcpp/time_source.cpp
Comment (jefferyyjhsu) [reply rclcpp/src/rclcpp/time_source.cpp:458]: I was going to move the check up, but c54a6f1cd2b5bfbe530c362ad7e8fa22f753e325 gave a good reasoning for why the check was removed from the old `on_set_parameters_callback` .

**Comment by iuhilnehc-ynos** [comment]
[CI](https://github.com/ros2/rclcpp/pull/1982#issuecomment-1257026005) seems to run with [the wrong settings](https://ci.ros2.org/job/ci_launcher/10878/parameters/).

**Comment by alsora** [comment]
My bad! I copied the repos file in the wrong field..!
New CI
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/17377/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/11911/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/17879/)

**Comment by alsora** [comment]
Hi, anything else to address before merging this PR?

**Comment by alsora** [comment]
@fujitatomoya @iuhilnehc-ynos is there anything else we should address before merging this PR?

**Comment by fujitatomoya** [comment]
@alsora some comments are not resolved or addressed yet?

I am okay to go with this, but requesting another maintainer to take a look at this, since it changes `Node` API and meaning of `NodeClock::ros_clock_`.

CC: @ivanpauno @clalancette @wjwwood

File: rclcpp/src/rclcpp/time_source.cpp
Comment (jefferyyjhsu) [reply rclcpp/src/rclcpp/time_source.cpp:139]: After thinking about this, I still think it's easier to maintain if we keep the logic for RCL_ROS_TIME clock in one block(from line 131 to 152), and the rest for other clock types in another block (from 154 to 156). What do you think, @alsora?

File: rclcpp/src/rclcpp/time_source.cpp
Comment (jefferyyjhsu) [reply rclcpp/src/rclcpp/time_source.cpp]: Removed the null check for node_parameters_ since other parts of the code don't usually perform such a check.

File: rclcpp/src/rclcpp/time_source.cpp
Comment (alsora) [reply rclcpp/src/rclcpp/time_source.cpp:139]: Good for me, this was just a minor comment

**Comment by jefferyyjhsu** [comment]
@fujitatomoya, I went back and resolved some of the minor comments that @alsora and I discussed internally and decided to keep it the same way. Please let me know if I am still missing anything. Thanks!

File: rclcpp/src/rclcpp/time_source.cpp
Comment (ivanpauno) [inline rclcpp/src/rclcpp/time_source.cpp:290]: these changes look unrelated to adding clock_type to node_options, right?

File: rclcpp/src/rclcpp/time_source.cpp
Comment (jefferyyjhsu) [reply rclcpp/src/rclcpp/time_source.cpp:131]: We actually check if the user is trying to set `ros time` to true and throw an exception in that case on line 154.

File: rclcpp/src/rclcpp/time_source.cpp
Comment (jefferyyjhsu) [reply rclcpp/src/rclcpp/time_source.cpp:290]: The callback is for handling illegal use_sim_time parameter set discussed [here](https://github.com/ros2/rclcpp/pull/1982#discussion_r978819552).

File: rclcpp/src/rclcpp/time_source.cpp
Comment (ivanpauno) [reply rclcpp/src/rclcpp/time_source.cpp:131]: Ok.
I think that failing earlier may make the logic a bit easier to read, but I don't mind.

File: rclcpp/src/rclcpp/time_source.cpp
Comment (ivanpauno) [inline rclcpp/src/rclcpp/time_source.cpp:463]: I think that `reason` should only be filled in case of failure, for the success case an empty string is fine.

**Review by ivanpauno — APPROVED** [review]
LGTM with minor comments addressed and green CI

**Comment by fujitatomoya** [comment]
@jefferyyjhsu thanks for iterating with us, final CI below.

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/17612/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/12154/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/18182/)

**Comment by fujitatomoya** [comment]
@jefferyyjhsu can you address DCO issue?

File: rclcpp/src/rclcpp/time_source.cpp
Comment (jefferyyjhsu) [reply rclcpp/src/rclcpp/time_source.cpp]: Thanks, just pushed a fix to this!

**Comment by jefferyyjhsu** [comment]
@fujitatomoya , yeah. Sorry, I did remember to fix it by rebasing it locally but forgot to finish it before pushing.

**Comment by alsora** [comment]
The windows failure happens also in a completely unrelated PR https://github.com/ros2/rclcpp/pull/1979

**Comment by fujitatomoya** [comment]
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/18368/)

**Comment by alsora** [comment]
Running full CI.
The github check reports failures on rclcpp tests but they seem unrelated.

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/17754/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/12296/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/18400/)

**Comment by alsora** [comment]
Ping; can we merge this PR?

**Comment by methylDragon** [comment]
Erm.. I'm getting the same issue locally as per the Rpr job
https://build.ros2.org/job/Rpr__rclcpp__ubuntu_jammy_amd64/469/testReport/junit/rclcpp/TestTimeSource/callbacks/

Something is causing the callback test to fail

**Comment by methylDragon** [comment]
Pinging @jefferyyjhsu

**Comment by methylDragon** [comment]
@ros-pull-request-builder retest this please!

**Comment by methylDragon** [comment]
I'm just gonna rerun the rpr job just in case: [Build Status](https://build.ros2.org/job/Rpr__rclcpp__ubuntu_jammy_amd64/501/)

**Comment by jefferyyjhsu** [comment]
@methylDragon, sorry for the delay. I tried reproducing the error with `colcon test --packages-select rclcpp` but failed. How are you running the tests?
In rclcpp/test/rclcpp/test_time_source.cpp, I only added 2 isolated test cases.

> TEST(TimeSource, valid_clock_type_for_sim_time)
> {
>   rclcpp::init(0, nullptr);
> 
>   rclcpp::NodeOptions options;
>   auto node = std::make_shared<rclcpp::Node>("my_node", options);
>   EXPECT_TRUE(
>     node->set_parameter(
>       rclcpp::Parameter(
>         "use_sim_time", rclcpp::ParameterValue(
>           true))).successful);
>   rclcpp::shutdown();
> }
> 
> TEST(TimeSource, invalid_clock_type_for_sim_time)
> {
>   rclcpp::init(0, nullptr);
> 
>   rclcpp::NodeOptions options;
>   options.clock_type(RCL_STEADY_TIME);
>   auto node = std::make_shared<rclcpp::Node>("my_node", options);
>   EXPECT_FALSE(
>     node->set_parameter(
>       rclcpp::Parameter(
>         "use_sim_time", rclcpp::ParameterValue(
>           true))).successful);
>   rclcpp::shutdown();
> }
>

**Comment by methylDragon** [comment]
> @methylDragon, sorry for the delay. I tried reproducing the error with `colcon test --packages-select rclcpp` but failed. How are you running the tests? In rclcpp/test/rclcpp/test_time_source.cpp, I only added 2 isolated test cases.
> 
> > TEST(TimeSource, valid_clock_type_for_sim_time)
> > {
> > rclcpp::init(0, nullptr);
> > rclcpp::NodeOptions options;
> > auto node = std::make_sharedrclcpp::Node("my_node", options);
> > EXPECT_TRUE(
> > node->set_parameter(
> > rclcpp::Parameter(
> > "use_sim_time", rclcpp::ParameterValue(
> > true))).successful);
> > rclcpp::shutdown();
> > }
> > TEST(TimeSource, invalid_clock_type_for_sim_time)
> > {
> > rclcpp::init(0, nullptr);
> > rclcpp::NodeOptions options;
> > options.clock_type(RCL_STEADY_TIME);
> > auto node = std::make_sharedrclcpp::Node("my_node", options);
> > EXPECT_FALSE(
> > node->set_parameter(
> > rclcpp::Parameter(
> > "use_sim_time", rclcpp::ParameterValue(
> > true))).successful);
> > rclcpp::shutdown();
> > }

Hey there! I am also running it with `colcon test --packages-select rclcpp`. It seems like this issue only happens randomly though, so I suspect it's a flaky test or a previously existing flaky test that's excacerbated by a change here? Or I could be wrong and its completely unrelated to this PR :o

**Comment by ivanpauno** [comment]
> Erm.. I'm getting the same issue locally as the Rpr job
https://build.ros2.org/job/Rpr__rclcpp__ubuntu_jammy_amd64/469/testReport/junit/rclcpp/TestTimeSource/callbacks/
>
> Something is causing the callback test to fail, though it happens sporadically

I think that test has been flaky for a long time (before this PR was merged), I don't think the failure is related to these changes.

**Comment by methylDragon** [comment]
Ah, okay, disregard my comment then! Sorry for the noise :bow:

**Pull request merged**
